src/mpid/ch4/src/ch4r_symheap.h

```diff
@@ -222,71 +235,99 @@ static inline int MPIDI_CH4I_allocate_symshm_segment(MPIR_Comm * shm_comm_ptr,
+
+            if (all_map_result != MPIDI_CH4I_SYMSHM_SUCCESS)
+                goto map_fail;
+        } else
+            goto map_fail;
     } else {
+        char serialized_hnd[MPL_SHM_GHND_SZ] = { 0 };
+
         /* receive the mapping result of rank 0 */
-        mpi_errno = MPIR_Bcast(mapfail_flag, 1, MPI_INT, 0, shm_comm_ptr, &errflag);
+        mpi_errno = MPIR_Bcast(&map_result, 1, MPI_INT, 0, shm_comm_ptr, &errflag);
+        if (map_result != MPIDI_CH4I_SYMSHM_SUCCESS) {
+            all_map_result = map_result;
+            goto map_fail;
+        }
 
         /* if rank 0 mapped successfully, others on the node attach shared memory region */
-        if (!*mapfail_flag) {
-            char serialized_hnd[MPL_SHM_GHND_SZ] = { 0 };
 
-            /* get serialized handle from rank 0 and deserialize it */
-            mpi_errno = MPIR_Bcast(serialized_hnd, MPL_SHM_GHND_SZ, MPI_CHAR, 0,
-                                   shm_comm_ptr, &errflag);
-            MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+        /* get serialized handle from rank 0 and deserialize it */
+        mpi_errno = MPIR_Bcast(serialized_hnd, MPL_SHM_GHND_SZ, MPI_CHAR, 0,
+                               shm_comm_ptr, &errflag);
+        MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
 
-            mpl_err = MPL_shm_hnd_deserialize(*shm_segment_hdl_ptr, serialized_hnd,
-                                              strlen(serialized_hnd));
-            MPIR_ERR_CHKANDJUMP(mpl_err != MPL_SHM_SUCCESS, mpi_errno, MPI_ERR_OTHER,
-                                "**alloc_shar_mem");
+        mpl_err = MPL_shm_hnd_deserialize(*shm_segment_hdl_ptr, serialized_hnd,
+                                          strlen(serialized_hnd));
+        if (mpl_err != MPL_SHM_SUCCESS) {
+            map_result = MPIDI_CH4I_SYMSHM_OTHER_FAIL;
+            goto result_sync;
+        }
 
-            /* try to attach with specified symmetric address */
-            mpl_err = MPL_shm_fixed_seg_attach(*shm_segment_hdl_ptr, shm_segment_len, base_ptr, 0);
-            if (mpl_err == MPL_SHM_EINVAL)
-                *mapfail_flag = 1;
+        /* try to attach with specified symmetric address */
+        mpl_err = MPL_shm_fixed_seg_attach(*shm_segment_hdl_ptr, shm_segment_len, base_ptr, 0);
+        if (mpl_err != MPL_SHM_SUCCESS) {
+            map_result = (mpl_err == MPL_SHM_EINVAL) ?
+                MPIDI_CH4I_SYMSHM_MAP_FAIL : MPIDI_CH4I_SYMSHM_OTHER_FAIL;
+            goto result_sync;
+        } else
+            mapped_flag = true;
 
-            /* handle mapping failure separately, caller may retry predefined times before giving up */
-            MPIR_ERR_CHKANDJUMP((mpl_err != MPL_SHM_EINVAL && mpl_err != MPL_SHM_SUCCESS),
-                                mpi_errno, MPI_ERR_OTHER, "**alloc_shar_mem");
+      result_sync:
+        /* check results of all processes, MAP_FAIL if no OTHER_FAIL is reported. */
```

Comment:
"MAP_FAIL if no OTHER_FAIL is reported" -- I don't understand what this means. The code looks like "MAP_FAIL if any process reports failure".

Reply:
I improved the comment. It should make it clear.
```
        /* check results of all processes. If any failure happens (max result > 0),
         * return MPIDI_CH4I_SYMSHM_OTHER_FAIL if anyone reports it (max result == 2).
         * Otherwise return MPIDI_CH4I_SYMSHM_MAP_FAIL (max result == 1). */
```